lib/spack/spack/solver/asp.py

```
@@ -1839,7 +1839,11 @@ def preferred_variants(self, pkg_name):
 
             # perform validation of the variant and values
             spec = spack.spec.Spec(pkg_name)
-            spec.update_variant_validate(variant_name, values)
+            try:
+                spec.update_variant_validate(variant_name, values)
+            except spack.error.SpackError:
```

Do we have a more specific exception for this? Like VariantValueError or whatever?

---

Also, what about requirements?

---

> Also, what about requirements?

Requirements are fine already. They will error, unless the requirement comes from the `all:` section in `packages.yaml`:

https://github.com/spack/spack/blob/868a3c43e4206f864121cdb81f9613ff260a3a3b/lib/spack/spack/solver/asp.py#L1710-L1715

The overall idea there is that, since requirements are stricter, we shouldn't silently skip them if they cannot be applied.[^1]

[^1]: Think of a typo like `+opnmp`. It would be better to error, than to skip and concretize to the opposite value.

---

Why guard against a typo `+opnmp` if `+openmp` errors because the majority of the specs don't have that variant. Makes little sense

---

I'll repeat. For requirements `all:` is permissive specific packages are not.

---

See comment below

---

Ok, tried it out, and we have a regression in the logic program :sweat: (so we don't raise here, but emit an unsat problem).